torch/csrc/jit/passes/memory_planning.cpp

```diff
@@ -0,0 +1,159 @@
+  TORCH_CHECK(
+      ttp->scalarType().has_value(),
+      "This output was profiled but didn't have a scalar type.")
+  TORCH_CHECK(
+      ttp->sizes().concrete_sizes().has_value(),
+      "This output was profiled but doesn't have sizes.")
+  return ttp;
+}
+
+std::pair<std::vector<int64_t>, std::vector<int64_t>> getSizesStrides(
+    const c10::TensorTypePtr& ttp) {
+  auto sizes = ttp->sizes().concrete_sizes().value();
+  std::vector<int64_t> strides;
+  auto _strides = ttp->strides().concrete_sizes();
+  if (_strides.has_value() && _strides.value()[0] != 0) {
+    strides = _strides.value();
+  } else {
+    strides = at::detail::defaultStrides(sizes);
+  }
+  return std::make_pair(sizes, strides);
+}
+
+size_t computeStorageSize(const c10::TensorTypePtr& ttp) {
+  std::vector<int64_t> sizes, strides;
+  std::tie(sizes, strides) = getSizesStrides(ttp);
+  return at::detail::computeStorageNbytes(
+      sizes, strides, elementSize(ttp->scalarType().value()));
+}
+
+std::pair<size_t, std::map<Value*, c10::TensorTypePtr>>
+computeAlignedStorageSize(std::vector<Node*>& nodes) {
+  // TODO: liveness analysis
+  std::map<Value*, c10::TensorTypePtr> managed_tensors;
+  size_t managed_bytes = 0;
+  for (const auto& node : nodes) {
+    for (const auto& output : node->outputs()) {
+      auto ttp = checkIsSupported(*output);
+      managed_bytes += computeStorageSize(ttp);
+      managed_tensors.insert({output, ttp});
+    }
+  }
+
+  auto aligned_size = MemoryPlanner::compute_aligned_tensor_size(managed_bytes);
+  return std::make_pair(aligned_size, managed_tensors);
+}
+
+DeviceType findDeviceType(std::shared_ptr<Graph>& graph) {
+  for (const auto& inp : graph->inputs()) {
+    auto typ = inp->type()->cast<TensorType>();
+    if (typ && typ->device().has_value()) {
+      return typ->device().value().type();
+    }
+  }
+
+  for (const auto& node : graph->nodes()) {
+    for (const auto& inp : node->inputs()) {
+      auto typ = inp->type()->cast<TensorType>();
+      if (typ && typ->device().has_value()) {
+        return typ->device().value().type();
+      }
+    }
+  }
+  TORCH_CHECK(false, "couldn't find device type")
+}
+
+Node* insertAllocStorageNode(std::shared_ptr<Graph>& graph, size_t total_size) {
+  auto deviceType = findDeviceType(graph);
+  auto* storage = graph->create(prim::AllocateStorage, 1);
+  // TODO: pass device type here
+  storage->i_(attr::total_size, total_size);
+  storage->i_(attr::device, static_cast<int8_t>(deviceType));
+  storage->insertBefore(graph->nodes().front());
+  return storage;
+}
+
+void insertAllocTensorNodes(
+    std::shared_ptr<Graph>& graph,
+    std::vector<Node*> out_nodes,
+    Node* storage,
+    std::map<Value*, c10::TensorTypePtr>& managed_tensors) {
+  size_t offset = 0;
+  size_t total_size = storage->i(attr::total_size);
+  for (auto* node : out_nodes) {
+    auto* alloc = graph->create(prim::AllocateTensor, 1);
+    node->addInput(alloc->output());
```

Comment:
looks like we are turning an op here into its out variant by adding an extra input? This is probably worth a comment or two.

Reply:
this is @eellison's insight - that nodes don't own schema but actually look up schema in the registry based number and type of inputs/outputs. will add a comment for sure because it was surprising to me as well.

Reply:
exactly, it's not super obvious :)